src/ergw_tei_mngr.erl

```diff
@@ -0,0 +1,170 @@
+    ets:delete(TID, KeyFun(TEI)),
+    Actions = [{reply, From, ok}, state_timeout()],
+    {keep_state_and_data, Actions};
+
+handle_event({call, From}, Request, _State, _Data) ->
+    Reply = {error, {badarg, Request}},
+    {keep_state_and_data, [{reply, From, Reply}]};
+
+handle_event(_, _Msg, _State, _Data) ->
+    keep_state_and_data.
+
+terminate(_Reason, _State, _Data) ->
+    ok.
+
+code_change(_OldVsn, State, Data, _Extra) ->
+    {ok, State, Data}.
+
+%%%=========================================================================
+%%%  internal functions
+%%%=========================================================================
+
+maybe_init_rnd(Name, State) when is_map_key(Name, State) ->
+    maps:get(Name, State);
+maybe_init_rnd(_, _) ->
+    rand:seed_s(exrop).
+
+state_timeout() ->
+    {state_timeout, ?SAVE_TIMEOUT, save}.
+
+serialize_data(State, #data{tid = TID}) ->
+    SeedL = maps:fold(fun(K,V,L) -> [{K, rand:export_seed_s(V)}|L] end, [], State),
+    Term =
+	#{seed  => SeedL,
+	  teids => ets:tab2list(TID)},
+    term_to_binary(Term, [{minor_version, 2}, compressed]).
+
+save_data(State, #data{node_id = NodeId, block_id = BlockId} = Data) ->
+    Serialized = serialize_data(State, Data).
+
+alloc_tei_reply(From, Reply, Name, RndState, State, Data) ->
+    Actions = [{reply, From, Reply}, state_timeout()],
+    {next_state, State#{Name => RndState}, Data, Actions}.
+
+alloc_tei(From, Name, _KeyFun, 0, RndState, State, Data) ->
+    alloc_tei_reply(From, {error, no_tei}, Name, RndState, State, Data);
+alloc_tei(From, Name, KeyFun, Cnt, RndState0, State,
+	  #data{prefix = Prefix0, prefix_len = PrefixLen, tid = TID} = Data) ->
+    Prefix = Prefix0 bsl (32 - PrefixLen),
+    Mask = 16#ffffffff bsr PrefixLen,
+    {TEI0, RndState} = rand:uniform_s(16#fffffffe, RndState0),
+    TEI = Prefix + (TEI0 band Mask),
+    case ets:insert_new(TID, {KeyFun(TEI), Name}) of
+	true ->
+	    alloc_tei_reply(From, {ok, TEI}, Name, RndState, State, Data);
+	false ->
+	    alloc_tei(From, Name, KeyFun, Cnt - 1, RndState, State, Data)
+    end.
+
+get_teid_config() ->
+	get_teid_pref_len(application:get_env(ergw, teid), {1, 0}).
+
+get_teid_pref_len({ok, {Prefix, Len} = WrongConfig}, DefaultConfig)
+	when not is_integer(Len); Len < 0; Len > ?MAX_TRIES; not is_integer(Prefix); Prefix < 0; Prefix > 8 ->
+		?LOG(warning, "TEID length and/or prefix has wrong configuration: ~p", [WrongConfig]),
+		DefaultConfig;
+get_teid_pref_len({ok, {Prefix, Len}}, _) ->
+	{Prefix, Len};
+get_teid_pref_len({ok, {Len}}, DefaultConfig) when not is_integer(Len); Len < 0; Len > ?MAX_TRIES ->
+	?LOG(warning, "TEID length has wrong configuration: ~p", [Len]),
+	DefaultConfig;
+get_teid_pref_len({ok, {Len}}, _) ->
+	{(fun(X) when X =< 8 -> X; (_) -> 1 end)(?POW(2, Len)), Len};
+get_teid_pref_len(WrongConfig, DefaultConfig) ->
+	?LOG(warning, "TEID has wrong configuration: ~p", [WrongConfig]),
+	DefaultConfig.
```

Comment:
please look how the other modules implement validate_config methods or implement this in ergw_config.